Sources/Foundation/JSONEncoder.swift

```diff
@@ -237,176 +237,405 @@ open class JSONEncoder {
+private enum JSONFuture {
+    case value(JSONValue)
+    case encoder(JSONEncoderImpl)
+    case nestedArray(Array)
+    case nestedObject(Object)
+    
+    class Array {
+        private(set) var array: [JSONFuture] = []
+
+        init() {
+            self.array.reserveCapacity(10)
+        }
+
+        @inline(__always) func append(_ element: JSONValue) {
+            self.array.append(.value(element))
+        }
+        
+        @inline(__always) func append(_ encoder: JSONEncoderImpl) {
+            self.array.append(.encoder(encoder))
+        }
+
+        @inline(__always) func appendArray() -> Array {
+            let array = Array()
+            self.array.append(.nestedArray(array))
+            return array
+        }
+
+        @inline(__always) func appendObject() -> Object {
+            let object = Object()
+            self.array.append(.nestedObject(object))
+            return object
+        }
+
+        var values: [JSONValue] {
+            self.array.map { (future) -> JSONValue in
+                switch future {
+                case .value(let value):
+                    return value
+                case .nestedArray(let array):
+                    return .array(array.values)
+                case .nestedObject(let object):
+                    return .object(object.values)
+                case .encoder(let encoder):
+                    return encoder.value ?? .object([:])
+                }
+            }
         }
     }
-}
 
-// MARK: - _JSONEncoder
+    class Object {
+        private(set) var dict: [String: JSONFuture] = [:]
 
-fileprivate class _JSONEncoder : Encoder {
-    // MARK: Properties
+        init() {
+            self.dict.reserveCapacity(20)
+        }
 
-    /// The encoder's storage.
-    fileprivate var storage: _JSONEncodingStorage
+        @inline(__always) func set(_ value: JSONValue, for key: String) {
+            self.dict[key] = .value(value)
+        }
 
-    /// Options set on the top-level encoder.
-    fileprivate let options: JSONEncoder._Options
+        @inline(__always) func setArray(for key: String) -> Array {
+            switch self.dict[key] {
+            case .encoder:
+                preconditionFailure("For key \"\(key)\" an encoder has already been created.")
+            case .nestedObject:
+                preconditionFailure("For key \"\(key)\" a keyed container has already been created.")
+            case .nestedArray(let array):
+                return array
+            case .none, .value:
+                let array = Array()
+                dict[key] = .nestedArray(array)
+                return array
+            }
+        }
 
-    /// The path to the current point in encoding.
-    public var codingPath: [CodingKey]
+        @inline(__always) func setObject(for key: String) -> Object {
+            switch self.dict[key] {
+            case .encoder:
+                preconditionFailure("For key \"\(key)\" an encoder has already been created.")
```

Comment:
is it safe to precondition here? i.e no user action can cause this?

Reply:
This shall lead to a crash because of a programmer error and closely mimics the Darwin Foundation behavior. It is not possible to create a json array or a json dictionary at the same time for the same key.